Cython/Compiler/ExprNodes.py

```diff
@@ -11364,13 +11364,13 @@ def compute_c_result_type(self, type1, type2):
                 self, type1, type2)
 
     def py_operation_function(self, code):
-        is_unicode_concat = False
-        if isinstance(self.operand1, FormattedValueNode) or isinstance(self.operand2, FormattedValueNode):
-            is_unicode_concat = True
-        else:
-            type1, type2 = self.operand1.type, self.operand2.type
-            if type1 is unicode_type or type2 is unicode_type:
-                is_unicode_concat = type1.is_builtin_type and type2.is_builtin_type
+        language_level = code.funcstate.scope.global_scope().context.language_level
+        type1, type2 = self.operand1.type, self.operand2.type
+        # in Py2 str_type is bytes but this can be concatted to a unicode_type string.
+        # in Py3 str_type is unicode so can be concatted to unicode_type or str_type
+        is_unicode_concat = (type1 in (unicode_type, str_type) and type2 in ((unicode_type, str_type))
+                                and not (type1 is str_type and type2 is str_type
+                                         and language_level == 2))
```

Comment:
Is there a reason why we need to look at the `language_level` here?
ISTM that all we need is 1) one of `type1` and `type2` is `unicode_type`, and 2) the other is either `unicode_type` or `str_type`.